src/safeposix/syscalls/net_calls.rs

```diff
@@ -1234,4 +1233,274 @@ impl Cage {
+        }
+        return Self::_epoll_object_allocator(self);
+    }
+
+    //this one can still be optimized
+    pub fn epoll_ctl_syscall(&self, epfd: i32, op: i32, fd: i32, event: EpollEvent) -> i32 {
+
+        let mut fdtable = self.filedescriptortable.write().unwrap();
+
+        //making sure that the epfd is really an epoll fd
+        if let Some(wrappedfd) = fdtable.get(&epfd) {
+            let mut filedesc_enum_epollfd = wrappedfd.write().unwrap();
+            if let Epoll(epollfdobj) = &mut *filedesc_enum_epollfd {
+
+                //check if the other fd is an epoll or not...
+                if let Epoll(_) = &*fdtable.get(&fd).unwrap().read().unwrap() {
+                    return syscall_error(Errno::EBADF, "epoll ctl", "provided fd is not a valid file descriptor")
+                }
+
+                //now that we know that the types are all good...
+                match op {
+                    EPOLL_CTL_DEL => {
+                        //since remove returns the value at the key and the values will always be EpollEvents, 
+                        //I am using this to optimize the code
+                        if let Some(EpollEvent{ events, fd }) = epollfdobj.registered_fds.remove(&fd) {} else {
+                            return syscall_error(Errno::ENOENT, "epoll ctl", "fd is not registered with this epfd");
+                        }
+                    }
+                    EPOLL_CTL_MOD => {
+                        ///check if the fd that we are modifying exists or not
+                        if let Some(_) = epollfdobj.registered_fds.get(&fd) {} else {
+                            return syscall_error(Errno::ENOENT, "epoll ctl", "fd is not registered with this epfd");
+                        }
+                        //if the fd already exists, insert overwrites the prev entry
+                        epollfdobj.registered_fds.insert(fd, EpollEvent { events: event.events, fd: event.fd });
+                    }
+                    EPOLL_CTL_ADD => {
+                        if let Some(_) = epollfdobj.registered_fds.get(&fd) {
+                            return syscall_error(Errno::EEXIST, "epoll ctl", "fd is already registered");
+                        }
+                        epollfdobj.registered_fds.insert(fd, EpollEvent { events: event.events, fd: event.fd });
+                    }
+                    _ => {
+                        return syscall_error(Errno::EINVAL, "epoll ctl", "provided op is invalid");
+                    }
+                }
+            } else {
+                return syscall_error(Errno::EBADF, "epoll ctl", "provided fd is not a valid file descriptor")
+            }
+        } else {
+            return syscall_error(Errno::EBADF, "epoll ctl", "provided epoll fd is not a valid epoll file descriptor");
+        }
+        return 0;
+    }
+
+    pub fn epoll_wait_syscall(&self, epfd: i32, maxevents: usize, timeout: Option<interface::RustDuration>) -> i32 {
```

Comment:
For epoll_wait, take a look at the man page for how to return things. Currently you just populate an array that nothing happens to. What you should do is take in a mutable slice of `struct epoll_event`s that is maxevents long and populate up to maxevents of them.